src/TesApi.Web/Storage/TerraStorageAccessProvider.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using TesApi.Web.Management.Clients;
+using TesApi.Web.Management.Configuration;
+using TesApi.Web.Management.Models.Terra;
+
+namespace TesApi.Web.Storage
+{
+    /// <summary>
+    /// Provides methods for blob storage access by using local path references in form of /storageaccount/container/blobpath
+    /// </summary>
+    public class TerraStorageAccessProvider : StorageAccessProvider
+    {
+        private readonly TerraOptions terraOptions;
+        private readonly TerraWsmApiClient terraWsmApiClient;
+        private const int FirstSegment = 0;
+        private const int SecondSegment = 1;
+        private const string SasTokenPermissions = "racw";
+
+        /// <summary>
+        /// Provides methods for blob storage access by using local path references in form of /storageaccount/container/blobpath
+        /// for Terra
+        /// </summary>
+        /// <param name="logger">Logger <see cref="ILogger"/></param>
+        /// <param name="terraOptions"><see cref="TerraOptions"/></param>
+        /// <param name="azureProxy">Azure proxy <see cref="IAzureProxy"/></param>
+        /// <param name="terraWsmApiClient"><see cref="terraWsmApiClient"/></param>
+        public TerraStorageAccessProvider(ILogger<TerraStorageAccessProvider> logger,
+            IOptions<TerraOptions> terraOptions, IAzureProxy azureProxy, TerraWsmApiClient terraWsmApiClient) : base(
+            logger, azureProxy)
+        {
+            this.terraWsmApiClient = terraWsmApiClient;
+            ArgumentNullException.ThrowIfNull(terraOptions);
+
+            this.terraOptions = terraOptions.Value;
+
+        }
+
+        /// <inheritdoc />
+        public override Task<bool> IsPublicHttpUrlAsync(string uriString)
```

Comment:
@BMurri and @MattMcL4475, the assumption for Terra is that all Http Urls will be public endpoints. Is this the best way to implement the assumption - is there any other case we need to consider?

Reply:
`IsPublicHttpUrlAsync` means that it can be reasonably assumed that the URL will not require any appended SAS (either because it already has one or because it points to an azure storage account and TES does not believe that it has access to request a SAS from that account.


If you have a URL for which you have a reasonable expectation to have to obtain a SAS token, then you should return `false` and expect to receive a callback to get a token.